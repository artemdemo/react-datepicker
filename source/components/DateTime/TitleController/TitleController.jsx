import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { propIsMoment } from '../propTypes';

import './TitleController.less';

class TitleController extends React.Component {
    changeForward() {
        const { onChangeForward } = this.props;
        onChangeForward();
    }

    changeBackward() {
        const { onChangeBackward } = this.props;
        onChangeBackward();
    }

    handleTitleClick() {
        const { onTitleClick } = this.props;
        onTitleClick && onTitleClick();
    }

    render() {
        const { date, width, format, onTitleClick } = this.props;
        const titleClass = classnames({
            'datetime-title-controller-title': true,
            'datetime-title-controller-title_clickable': onTitleClick,
        });
        return (
            <table
                className='datetime-title-controller'
                style={{
                    width: width || 'auto',
                }}
            >
                <thead>
                    <tr>
                        <th
                            className='datetime-title-controller-change
                                       datetime-title-controller-change_backward'
                            onClick={this.changeBackward.bind(this)}
                        />
                        <th
                            className={titleClass}
                            colSpan='5'
                            onClick={this.handleTitleClick.bind(this)}
                        >
                            {date.format(format)}
                        </th>
                        <th
                            className='datetime-title-controller-change
                                       datetime-title-controller-change_forward'
                            onClick={this.changeForward.bind(this)}
                        />
                    </tr>
                </thead>
            </table>
        );
    }
}

TitleController.displayName = 'TitleController';

TitleController.propTypes = {
    date: propIsMoment.isRequired,
    onChangeForward: PropTypes.func.isRequired,
    onChangeBackward: PropTypes.func.isRequired,
    format: PropTypes.string.isRequired,
    width: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
    ]),
    onTitleClick: PropTypes.func,
};

TitleController.defaultProps = {
    onTitleClick: null,
    width: 'auto',
};

export default TitleController;
