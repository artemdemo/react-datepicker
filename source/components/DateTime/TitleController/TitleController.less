@import '../styles/general';

.datetime-title-controller {
    .datetime-table();
}

.datetime-title-controller-change {
    .clickable();

    padding: 15px;
    position: relative;

    &::after {
        position: absolute;
        top: 5%;
        left: 43%;
    }
}
    .datetime-title-controller-change_backward {
        &::after {
            content: '‹';
        }
    }
    .datetime-title-controller-change_forward {
        &::after {
            content: '›';
        }
    }

.datetime-title-controller-title {
    font-weight: normal;
}
    .datetime-title-controller-title_clickable {
        .clickable();
    }
