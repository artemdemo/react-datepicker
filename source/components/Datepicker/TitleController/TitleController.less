@import '../datepicker-general';

.datepicker-title-controller {
    .datepicker-table();
}

.datepicker-title-controller-change-month {
    cursor: pointer;
    padding: 15px;
    border-radius: 4px;
    position: relative;

    &:hover {
        background-color: fade(@hover-color, 50%);
    }
    &::after {
        position: absolute;
        top: 20%;
        left: 43%;
    }
}
    .datepicker-title-controller-change-month_backward {
        &::after {
            content: '‹';
        }
    }
    .datepicker-title-controller-change-month_forward {
        &::after {
            content: '›';
        }
    }

.datepicker-title-controller-current-month-year {
    font-weight: normal;
}
