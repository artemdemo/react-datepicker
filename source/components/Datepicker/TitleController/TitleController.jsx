import React from 'react';
import PropTypes from 'prop-types';

import './TitleController.less';

class TitleController extends React.Component {
    changeMonthForward() {
        const { onChangeMonth, date } = this.props;
        onChangeMonth && onChangeMonth(date.clone().add(1, 'month'));
    }

    changeMonthBackward() {
        const { onChangeMonth, date } = this.props;
        onChangeMonth && onChangeMonth(date.clone().subtract(1, 'month'));
    }

    render() {
        const { date } = this.props;
        return (
            <table className='datepicker-title-controller'>
                <thead>
                    <tr>
                        <th
                            className='datepicker-title-controller-change-month
                                       datepicker-title-controller-change-month_backward'
                            onClick={this.changeMonthBackward.bind(this)} />
                        <th
                            className='datepicker-title-controller-current-month-year'
                            colSpan='5'>
                            {date.format('MMMM, YYYY')}
                        </th>
                        <th
                            className='datepicker-title-controller-change-month
                                       datepicker-title-controller-change-month_forward'
                            onClick={this.changeMonthForward.bind(this)} />
                    </tr>
                </thead>
            </table>
        );
    }
}

TitleController.propTypes = {
    date: PropTypes.shape({}).isRequired,
    onChangeMonth: PropTypes.func,
};

TitleController.defaultProps = {
    onChangeMonth: null,
};

export default TitleController;
