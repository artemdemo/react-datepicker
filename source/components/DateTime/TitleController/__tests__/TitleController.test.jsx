import React from 'react';
import moment from 'moment';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import TitleController from '../TitleController';

describe('<TitleController />', () => {
    const date = moment(1502459094000);

    it('simple render', () => {
        const tree = renderer.create(
            <TitleController
                date={date}
                format='MMMM, YYYY'
                onChangeForward={() => {}}
                onChangeBackward={() => {}}
            />,
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('title clicked', () => {
        let titleClicked = false;
        const wrapper = mount(
            <TitleController
                date={date}
                format='MMMM, YYYY'
                onChangeForward={() => {}}
                onChangeBackward={() => {}}
                onTitleClick={() => titleClicked = true}
            />,
        );
        wrapper.find('.datetime-title-controller-title').first().simulate('click');
        expect(titleClicked).toBe(true);
    });

    it('backward clicked', () => {
        let clicked = false;
        const wrapper = mount(
            <TitleController
                date={date}
                format='MMMM, YYYY'
                onChangeForward={() => {}}
                onChangeBackward={() => clicked = true}
            />,
        );
        wrapper.find('.datetime-title-controller-change_backward').first().simulate('click');
        expect(clicked).toBe(true);
    });

    it('forward clicked', () => {
        let clicked = false;
        const wrapper = mount(
            <TitleController
                date={date}
                format='MMMM, YYYY'
                onChangeForward={() => clicked = true}
                onChangeBackward={() => {}}
            />,
        );
        wrapper.find('.datetime-title-controller-change_forward').first().simulate('click');
        expect(clicked).toBe(true);
    });
});
