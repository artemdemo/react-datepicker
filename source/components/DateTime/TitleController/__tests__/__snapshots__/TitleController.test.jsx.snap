// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TitleController /> simple render 1`] = `
<table
  className="datetime-title-controller"
  style={
    Object {
      "width": "auto",
    }
  }
>
  <thead>
    <tr>
      <th
        className="datetime-title-controller-change datetime-title-controller-change_backward"
        onClick={[Function]}
      />
      <th
        className="datetime-title-controller-title"
        colSpan="5"
        onClick={[Function]}
      >
        August, 2017
      </th>
      <th
        className="datetime-title-controller-change datetime-title-controller-change_forward"
        onClick={[Function]}
      />
    </tr>
  </thead>
</table>
`;
